File: sdks/python/src/opik/api_objects/prompt/prompt_cache.py
import atexit
import collections
import logging
import threading
import time
import typing

from opik import config as opik_config
from .base_prompt import BasePrompt

logger = logging.getLogger(__name__)

_MIN_REFRESH_INTERVAL_SECONDS = 1.0

_CacheKey = typing.Tuple[
    str,
    typing.Optional[str],
    typing.Optional[str],
    str,
    typing.Optional[str],
    typing.Optional[str],
]
#             name  pin(commit/ver)  project_name  tmpl  environment  mask_id

_RefreshCallback = typing.Callable[[], typing.Optional[BasePrompt]]

_MAX_CACHE_SIZE = 128


class _CachedPrompt:
    __slots__ = ("prompt", "ttl_seconds", "last_fetch", "refresh_callback")

    def __init__(
        self,
        prompt: BasePrompt,
        ttl_seconds: typing.Optional[int],
        refresh_callback: typing.Optional[_RefreshCallback] = None,
    ) -> None:
        self.prompt = prompt
        self.ttl_seconds = ttl_seconds
        self.last_fetch = time.monotonic()
        self.refresh_callback = refresh_callback

    def is_stale(self) -> bool:
        if self.ttl_seconds is None:
            return False
        return (time.monotonic() - self.last_fetch) >= self.ttl_seconds


class PromptCache:
    def __init__(self, max_size: int = _MAX_CACHE_SIZE) -> None:
        self._lock = threading.Lock()
        self._entries: collections.OrderedDict[_CacheKey, _CachedPrompt] = (
            collections.OrderedDict()
        )
        self._max_size = max_size
        self._thread: typing.Optional[threading.Thread] = None
        self._stop_event = threading.Event()

    def get(self, key: _CacheKey) -> typing.Optional[BasePrompt]:
        with self._lock:
            entry = self._entries.get(key)
            if entry is None:
                return None
            self._entries.move_to_end(key)
            return entry.prompt

    def clear(self) -> None:
        self._stop_refresh_thread()
        with self._lock:
            self._entries.clear()

    def invalidate_for_prompt(
        self, name: str, project_name: typing.Optional[str]
    ) -> None:
        """Drop every cached entry for the given prompt name + project scope.

        Used after operations that change the env-to-version mapping (such as
        ``Opik.set_prompt_environment``) so that subsequent
        ``get_prompt(..., environment=...)`` calls cannot return a stale version.
        """
        with self._lock:
            stale_keys = [
                key
                for key in self._entries
                if key[0] == name and key[2] == project_name
            ]
            for key in stale_keys:
                del self._entries[key]

    def get_or_fetch(
        self,
        key: _CacheKey,
        fetch_fn: _RefreshCallback,
        ttl_seconds: typing.Optional[int],
        refresh_callback: typing.Optional[_RefreshCallback] = None,
    ) -> typing.Optional[BasePrompt]:
        with self._lock:
            entry = self._entries.get(key)
            if entry is not None:
                self._entries.move_to_end(key)
                return entry.prompt

        prompt = fetch_fn()
        if prompt is None:
            return None

        cached = _CachedPrompt(
            prompt=prompt,
            ttl_seconds=ttl_seconds,
            refresh_callback=refresh_callback,
        )
        with self._lock:
            self._entries[key] = cached
            self._entries.move_to_end(key)
            while len(self._entries) > self._max_size:
                self._entries.popitem(last=False)

        if ttl_seconds is not None:
            self._ensure_refresh_thread_started()

        return prompt

    def _ensure_refresh_thread_started(self) -> None:
        with self._lock:
            if self._thread is not None and self._thread.is_alive():
                return
            self._stop_event.clear()
            self._thread = threading.Thread(
                target=self._refresh_loop, daemon=True, name="OpikPromptCacheRefresh"
            )
            self._thread.start()
            atexit.register(self._stop_refresh_thread)

    def _refresh_loop(self) -> None:
        while not self._stop_event.is_set():
            self._refresh_stale_entries()
            ttl = opik_config.OpikConfig().prompt_cache_ttl_seconds
            interval = max(float(ttl), _MIN_REFRESH_INTERVAL_SECONDS)
            self._stop_event.wait(interval)

    def _refresh_stale_entries(self) -> None:
        with self._lock:
            entries = list(self._entries.values())
        for entry in entries:
            if self._stop_event.is_set():
                break
            if not entry.is_stale() or entry.refresh_callback is None:
                continue
            try:
                new_prompt = entry.refresh_callback()
                if new_prompt is not None:
                    with self._lock:
                        entry.prompt = new_prompt
                        entry.last_fetch = time.monotonic()
            except Exception:
                logger.debug(
                    "Background prompt cache refresh failed for prompt %r",
                    entry.prompt.name,
                    exc_info=True,
                )

    def _stop_refresh_thread(self) -> None:
        with self._lock:
            if self._thread is None:
                return
            thread = self._thread
            self._stop_event.set()
            self._thread = None
        try:
            thread.join(timeout=5)
            if thread.is_alive():
                logger.error(
                    "Prompt cache refresh thread did not stop within the timeout."
                )
        except Exception:
            logger.exception(
                "Unexpected error while waiting for prompt cache refresh thread to stop."
            )


_cache = PromptCache()


def get_global_cache() -> PromptCache:
    return _cache


_PromptT = typing.TypeVar("_PromptT", bound=BasePrompt)


def get_or_fetch(
    name: str,
    commit: typing.Optional[str],
    project_name: typing.Optional[str],
    template_structure: str,
    fetch_fn: typing.Callable[[], typing.Optional[_PromptT]],
    mask_id: typing.Optional[str] = None,
    version: typing.Optional[str] = None,
    environment: typing.Optional[str] = None,
) -> typing.Optional[_PromptT]:
    """Return a cached prompt or fetch, cache, and return it.

    If the prompt is already cached, returns it immediately.
    Otherwise calls *fetch_fn* to produce one. If *fetch_fn* returns None
    (prompt not found), returns None without caching.

    ``commit`` and ``version`` reuse the same cache key slot (commits are 8
    hex chars, versions match ``v\\d+``, so they cannot collide). Only
    ``commit`` pins indefinitely — a sequential ``version`` like ``"v3"``
    can be reassigned by the backend if the underlying version is deleted
    and recreated, so it is subject to the normal TTL refresh just like
    "latest" lookups.

    For masked prompts (mask_id is not None), the entry is TTL-evicted like
    normal unpinned entries but has no background refresh callback.
    """
    identifier = version if version is not None else commit
    ttl = opik_config.OpikConfig().prompt_cache_ttl_seconds
    ttl_seconds = None if commit is not None else ttl
    refresh_callback = (
        fetch_fn if (ttl_seconds is not None and mask_id is None) else None
    )
    key: _CacheKey = (
        name,
        identifier,
        project_name,
        template_structure,
        environment,
        mask_id,
    )
    result = _cache.get_or_fetch(
        key=key,
        fetch_fn=fetch_fn,
        ttl_seconds=ttl_seconds,
        refresh_callback=refresh_callback,
    )
    return typing.cast(typing.Optional[_PromptT], result)


def invalidate_for_prompt(name: str, project_name: typing.Optional[str]) -> None:
    """Drop every cached entry for the given prompt name + project scope."""
    _cache.invalidate_for_prompt(name=name, project_name=project_name)

File: sdks/typescript/src/opik/prompt/promptCache.ts
import { logger } from "@/utils/logger";
import type { BasePrompt } from "./BasePrompt";

export const DEFAULT_TTL_SECONDS = 300;
const MIN_REFRESH_INTERVAL_SECONDS = 1.0;
const MAX_CACHE_SIZE = 128;
const MAX_CONCURRENT_REFRESHES = 100;

type RefreshCallback = () => Promise<BasePrompt | null>;

interface CachedPrompt {
  prompt: BasePrompt;
  ttlSeconds: number | null;
  lastFetch: number;
  refreshCallback?: RefreshCallback;

  get isStale(): boolean;
}

function makeCachedPrompt(
  prompt: BasePrompt,
  ttlSeconds: number | null,
  refreshCallback?: RefreshCallback
): CachedPrompt {
  return {
    prompt,
    ttlSeconds,
    lastFetch: Date.now(),
    refreshCallback,
    get isStale() {
      if (this.ttlSeconds === null) return false;
      return Date.now() - this.lastFetch >= this.ttlSeconds * 1000;
    },
  };
}

export class PromptCache {
  private entries = new Map<string, CachedPrompt>();
  private inflight = new Map<string, Promise<BasePrompt | null>>();
  private maxSize: number;
  private refreshTimer: ReturnType<typeof setInterval> | null = null;
  private refreshRunning = false;
  private stopped = false;

  constructor(maxSize = MAX_CACHE_SIZE) {
    this.maxSize = maxSize;
  }

  get(key: string): BasePrompt | null {
    const entry = this.entries.get(key);
    if (!entry) return null;
    // LRU: delete and re-insert to move to end
    this.entries.delete(key);
    this.entries.set(key, entry);
    return entry.prompt;
  }

  async getOrFetch(
    key: string,
    fetchFn: RefreshCallback,
    ttlSeconds: number | null,
    refreshCallback?: RefreshCallback
  ): Promise<BasePrompt | null> {
    const existing = this.entries.get(key);
    if (existing) {
      this.entries.delete(key);
      this.entries.set(key, existing);
      return existing.prompt;
    }

    const pending = this.inflight.get(key);
    if (pending) return pending;

    const promise = this.fetchAndCache(key, fetchFn, ttlSeconds, refreshCallback);
    this.inflight.set(key, promise);
    try {
      return await promise;
    } finally {
      this.inflight.delete(key);
    }
  }

  private async fetchAndCache(
    key: string,
    fetchFn: RefreshCallback,
    ttlSeconds: number | null,
    refreshCallback?: RefreshCallback
  ): Promise<BasePrompt | null> {
    const prompt = await fetchFn();
    if (prompt === null) return null;

    this.entries.set(
      key,
      makeCachedPrompt(
        prompt,
        ttlSeconds,
        refreshCallback,
      ),
    );
    this.evict();

    if (ttlSeconds !== null) {
      this.ensureRefreshTimerStarted(ttlSeconds);
    }

    return prompt;
  }

  evictByIds(ids: string[]): void {
    const idSet = new Set(ids);
    for (const [key, entry] of this.entries) {
      if (entry.prompt.id && idSet.has(entry.prompt.id)) {
        this.entries.delete(key);
      }
    }
  }

  /**
   * Drop every cached entry for the given prompt name + project scope.
   *
   * Used after operations that change the environment-to-version mapping (such
   * as `setPromptEnvironments` or assigning environments during `createPrompt`)
   * so that subsequent `getPrompt({ name, environment })` calls cannot return a
   * stale version.
   */
  invalidateForPrompt(name: string, projectName: string | undefined): void {
    const scope = projectName ?? "";
    for (const key of this.entries.keys()) {
      const [keyName, , keyProject] = JSON.parse(key) as string[];
      if (keyName === name && keyProject === scope) {
        this.entries.delete(key);
      }
    }
  }

  clear(): void {
    this.stopRefreshTimer();
    this.entries.clear();
    this.stopped = false;
  }

  private evict(): void {
    while (this.entries.size > this.maxSize) {
      const firstKey = this.entries.keys().next().value;
      if (firstKey !== undefined) {
        this.entries.delete(firstKey);
      }
    }
  }

  private ensureRefreshTimerStarted(ttlSeconds: number): void {
    if (this.stopped || this.refreshTimer !== null) return;

    const intervalMs =
      Math.max(ttlSeconds, MIN_REFRESH_INTERVAL_SECONDS) * 1000;

    this.refreshTimer = setInterval(() => {
      this.refreshStaleEntries().catch((err) => {
        logger.debug("Prompt cache background refresh loop failed", { error: err });
      });
    }, intervalMs);

    // Don't keep the Node.js process alive just for cache refresh
    if (this.refreshTimer && typeof this.refreshTimer === "object" && "unref" in this.refreshTimer) {
      this.refreshTimer.unref();
    }
  }

  private async refreshStaleEntries(): Promise<void> {
    if (this.refreshRunning) return;
    this.refreshRunning = true;
    try {
      const stale = Array.from(this.entries.values()).filter(
        (entry) => entry.isStale && entry.refreshCallback
      );

      for (let i = 0; i < stale.length; i += MAX_CONCURRENT_REFRESHES) {
        if (this.stopped) break;
        const batch = stale.slice(i, i + MAX_CONCURRENT_REFRESHES);
        const results = await Promise.allSettled(
          batch.map(async (entry) => {
            const newPrompt = await entry.refreshCallback!();
            if (newPrompt !== null) {
              entry.prompt = newPrompt;
              entry.lastFetch = Date.now();
            }
          })
        );

        for (let j = 0; j < results.length; j++) {
          const result = results[j];
          if (result.status === "rejected") {
            logger.debug("Background prompt cache refresh failed", {
              promptName: batch[j].prompt.name,
              error: result.reason,
            });
          }
        }
      }
    } finally {
      this.refreshRunning = false;
    }
  }

  private stopRefreshTimer(): void {
    this.stopped = true;
    if (this.refreshTimer !== null) {
      clearInterval(this.refreshTimer);
      this.refreshTimer = null;
    }
  }
}

const globalCache = new PromptCache();

export function getGlobalCache(): PromptCache {
  return globalCache;
}

export function buildCacheKey(
  name: string,
  commit: string | undefined,
  projectName: string | undefined,
  templateStructure: string,
  maskId?: string | null,
  version?: string,
  environment?: string,
): string {
  // version and commit can never collide (commits are 8 hex chars, versions are "v<N>")
  // so we reuse the same slot in the key.
  const pin = version ?? commit ?? "";
  return JSON.stringify([name, pin, projectName ?? "", templateStructure, environment ?? "", maskId ?? ""]);
}

export async function getOrFetch<T extends BasePrompt>(
  name: string,
  commit: string | undefined,
  projectName: string | undefined,
  templateStructure: string,
  fetchFn: () => Promise<T | null>,
  ttlSeconds?: number,
  maskId?: string | null,
  version?: string,
  environment?: string,
): Promise<T | null> {
  // Only commit pins indefinitely. A sequential version like "v3" can be
  // reassigned by the backend if the underlying version is deleted and
  // recreated, so it follows the normal TTL refresh.
  const resolvedTtl = commit != null ? null : (ttlSeconds ?? DEFAULT_TTL_SECONDS);
  const refreshCallback = resolvedTtl !== null && !maskId ? fetchFn : undefined;
  const key = buildCacheKey(name, commit, projectName, templateStructure, maskId, version, environment);
  const result = await globalCache.getOrFetch(key, fetchFn, resolvedTtl, refreshCallback);
  return result as T | null;
}
